fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/util/pair/ImmutablePair.java

```diff
@@ -0,0 +1,107 @@
+/*
+ * ImmutablePair.java
+ *
+ * This source file is part of the FoundationDB open source project
+ *
+ * Copyright 2015-2024 Apple Inc. and the FoundationDB project authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.apple.foundationdb.record.util.pair;
+
+import com.apple.foundationdb.annotation.API;
+
+import javax.annotation.Nonnull;
+import javax.annotation.Nullable;
+import java.util.Objects;
+
+/**
+ * Immutable implementation of the {@link Pair} interface.
+ *
+ * @param <L> the type of the left element of the pair
+ * @param <R> the type of the right element of the pair
+ */
+@API(API.Status.INTERNAL)
+public class ImmutablePair<L, R> implements Pair<L, R> {
+    @Nullable
+    private final L left;
+    @Nullable
+    private final R right;
+    private volatile int hashCode;
+
+    private ImmutablePair(@Nullable L left, @Nullable R right) {
+        this.left = left;
+        this.right = right;
+    }
+
+    @Nullable
+    @Override
+    public L getLeft() {
+        return left;
+    }
+
+    @Nullable
+    @Override
+    public R getRight() {
+        return right;
+    }
+
+    @Override
+    public R setValue(final R value) {
+        throw new UnsupportedOperationException("cannot set value on an immmutable Pair");
+    }
+
+    @Override
+    public boolean equals(final Object obj) {
+        if (this == obj) {
+            return true;
+        }
+        if (!(obj instanceof Pair<?, ?>)) {
+            return false;
+        }
+        Pair<?, ?> other = (Pair<?, ?>)obj;
+        return Objects.equals(getLeft(), other.getLeft()) && Objects.equals(getRight(), other.getRight());
+    }
+
+    @Override
+    public int hashCode() {
+        if (hashCode == 0) {
+            synchronized (this) {
+                if (hashCode == 0) {
+                    hashCode = Objects.hash(left, right);
+                }
```

Comment:
I do like this, do you think this is more efficient than using a lambda which lazily calculate (and cache) the hash code?
At first glance it seems like it is faster, but I am not sure considering that `serialized` could have an internal bound lock.

Reply:
One thing to note is that the most common technique that we use for memoizing things like this is using `Suppliers.memoize`, which has an internal double-checked lock very similar to this: https://github.com/google/guava/blob/858caf425cdf3183776b61b437ba7a4129b6539b/guava/src/com/google/common/base/Suppliers.java#L184

So, we wouldn't save the lock by switching to a lambda. We may save an allocation, though hotspot could be inlining some of that, so it's a bit hard to say. One approach that we could try here would be a non-locking double checked lock, that is, replacing this with something like:

```java
if (hashCode == 0) {
    hashCode = Objects.hash(left, right);
}
return hashCode;
```

As long as the `hashCode` variable is `volatile` and as long as the hashes of left and right always return the same value (which they should), I think this would work, and we'd avoid the lock at the cost of sometimes having multiple threads calculating the hash for the same `Pair`.

I'm not sure I have that much of a strong preference between this approach and using a (constant) lambda with `Suppliers.memoize`. I sort of think it's actually easier to see what's going on, but I grant that others may like delegating the lock to a library function.